### PR TITLE
docs: Add Docker section

### PR DESCRIPTION
**Pull request by hinxcode**
This PR is for adding the documentation section for Docker: [issue link](https://github.com/uw-ssec/rse-guidelines/issues/34)

I lifted the [original paragraph](https://rse-guidelines.readthedocs.io/en/latest/fundamentals/share-deploy.html#docker-containerization-for-consistency) written by @lsetiawan and added a subsection for [Docker Swarm](https://github.com/hinxcode/rse-guidelines/commit/c723852b889c558f7e3ffbfcc00d016ab452af75). As a result, the best way to review this PR is to review it commit by commit.

Please feel free to comment on this PR and let's make this doc better!

### CODE REVIEW

File: fundamentals/computing-development-environments/docker.md [hunk 1]
```diff
@@ -1,3 +1,127 @@
...
+
+Containerization has changed how we build and run applications. It allows apps to run consistently across different environments, from development laptops to production servers. However, as applications grow, managing many containers becomes challenging.
+
+This is where orchestration tools come in. They help automate the deployment, scaling, and management of containerized applications. Key benefits include:
+
+- Automating application maintenance
+- Automatically replacing failed containers
+- Managing updates and configurations
+
+Two popular orchestration tools are [Kubernetes](https://docs.docker.com/guides/deployment-orchestration/kube-deploy/) and [Swarm Mode](https://docs.docker.com/engine/swarm/).
```
Comment (nikiburggraf) [inline fundamentals/computing-development-environments/docker.md:79]: This part is, in my opinion, the perfect level of detail for this page. 

Maybe we add a few sentences about what k8s and Swarm mode are, too

File: fundamentals/computing-development-environments/docker.md [hunk 2]
```diff
@@ -1,3 +1,127 @@
...
+- Automatically replacing failed containers
+- Managing updates and configurations
+
+Two popular orchestration tools are [Kubernetes](https://docs.docker.com/guides/deployment-orchestration/kube-deploy/) and [Swarm Mode](https://docs.docker.com/engine/swarm/). 
+
+> **Note**  
+> There are some similar tools that you might find confusing when it comes to Docker Swarm:
+> 1. [Classic Swarm](https://github.com/docker-archive/classicswarm) (Classic and is no longer actively developed)
+> 2. [SwarmKit](https://github.com/moby/swarmkit)
+> 3. [Swarm mode](https://docs.docker.com/engine/swarm/) (Docker version >= 1.12)
```
Comment (nikiburggraf) [inline fundamentals/computing-development-environments/docker.md:85]: Put the item you're actually talking about as `1` and make it clear that that's the one you're referring to

For the other ones, a sentence like you have for `Classic Swarm` clarifying what the item is and why it isn't Swarm would be extremely helpful

eg


```suggestion
> **Note**  
> There are some similar tools that you might find confusing when it comes to Docker Swarm:
> 1. [Swarm mode](https://docs.docker.com/engine/swarm/) (Docker version >= 1.12)
>   * This is what we are referring to
> 2. [SwarmKit](https://github.com/moby/swarmkit) 
>   * This project contains the underlying libraries used by Swarm mode
> 3. [Classic Swarm](https://github.com/docker-archive/classicswarm) 
>   * Docker's original orchestration tool. Classic Swarm is no longer actively developed, and you should focus on Swarm mode instead
```

(Again please feel free to correct/reword/anything this suggestion)

**Review by nikiburggraf — CHANGES_REQUESTED** [review]
Overall, extremely solid work! 

Had a bunch of suggestions, and some concerns that details about Swarm mode might be better suited to the planned advanced section on Docker (see: https://github.com/uw-ssec/rse-guidelines/issues/60 )

File: fundamentals/computing-development-environments/docker.md (hunk 2 above)
Comment (hinxcode) [reply fundamentals/computing-development-environments/docker.md]: Done!

File: fundamentals/computing-development-environments/docker.md (hunk 1 above)
Comment (hinxcode) [reply fundamentals/computing-development-environments/docker.md]: Done!

File: fundamentals/computing-development-environments/docker.md [hunk 3]
```diff
@@ -1,3 +1,127 @@
...
+### 2. Docker Image
+
+A Docker Image is the "built" version of a Docker file.
+This typically live on your internal Docker host or an external Registry, so you can't store a Docker image in git.
+
+An example of a docker image that has been built with the Docker file above can be found [here](https://hub.docker.com/layers/pangeo/base-image/latest/images/sha256-1854ca62ef75f1e017e1920b3a167c62cb5f0ee921f13c3247fc4b016e5be3be?context=explore). That is the published image for with specific file sha, which allows for a specific version of that image, just like your regular files!
+
+### 3. Docker Container
+
+A container is a standard unit of software that get's spun up from an image.
```
[the remark this answers is not in the document]
Comment (hinxcode) [reply fundamentals/computing-development-environments/docker.md]: Fixed!

File: fundamentals/computing-development-environments/docker.md [hunk 4]
```diff
@@ -1,3 +1,127 @@
...
+To start using Docker Swarm mode:
+
+1. Initialize a swarm (cluster) of Docker Engines
+2. Add worker nodes to the swarm
+3. Deploy services to the swarm
+4. Scale services as needed
+
+For specific commands and a detailed tutorial, please refer to: [Swarm mode official tutorial](https://docs.docker.com/engine/swarm/swarm-tutorial/).
+
+### 5. Container Registry
```
[the remark this answers is not in the document]
Comment (hinxcode) [reply fundamentals/computing-development-environments/docker.md]: I totally agree with you. Done!

File: fundamentals/computing-development-environments/docker.md [hunk 5]
```diff
@@ -1,3 +1,127 @@
...
+
+As you saw and I mentioned above that images can be stored in a Registry.
+The most popular free registry currently is [Docker Hub](https://hub.docker.com/),
+but there is now the [Github Container Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry) as well.
+
+Similarly like an app store, the Docker Hub serves various images and image versions. As you can see [here](https://hub.docker.com/r/pangeo/base-image/tags) these are the various `tags` for the example image.
+Tags are like versions, you can tag images however you like!
+
+**You can go through the [Docker 101](https://www.docker.com/101-tutorial/) lab to practice the concepts above! And the [official docker website](https://docs.docker.com/get-started/) has lots of great "get started" learning resources.**
```
[the remark this answers is not in the document]
Comment (hinxcode) [reply fundamentals/computing-development-environments/docker.md]: Fixed!

File: fundamentals/computing-development-environments/docker.md [hunk 6]
```diff
@@ -1,3 +1,127 @@
...
+These containers can run consistently across different environments, eliminating the "it works on my machine" problem.
+
+**Key features**:
+
+- **Isolation**: Docker containers encapsulate applications and their dependencies, ensuring that they run consistently and independently of the host environment.
+- **Portability**: Docker containers can be easily moved between different environment: development, testing, and production. This guarantees consistent behavior.
+- **Version Control**: Docker file and/or images can be versioned, allowing you to maintain and track different versions of your application.
+- **Microservices Architecture**: Docker containers are well-suited for microservices-based architectures, enabling you to break down complex applications into smaller, manageable components.
+
+### Containers vs VM
```
[the remark this answers is not in the document]
Comment (hinxcode) [reply fundamentals/computing-development-environments/docker.md]: Fixed!

File: fundamentals/computing-development-environments/docker.md [hunk 7]
```diff
@@ -1,3 +1,127 @@
...
+
+![container-vm](https://www.docker.com/wp-content/uploads/2021/11/docker-containerized-and-vm-transparent-bg.png)
+
+### Virtual Machine (VM)
+
+Virtual machines (VMs) are an abstraction of physical hardware turning one server into many servers. The hypervisor allows multiple VMs to run on a single machine. Each VM includes a full copy of an operating system, the application, necessary binaries and libraries – taking up tens of GBs. VMs can also be slow to boot.
+
+### Containers
```
[the remark this answers is not in the document]
Comment (hinxcode) [reply fundamentals/computing-development-environments/docker.md]: Done!

File: fundamentals/computing-development-environments/docker.md [hunk 8]
```diff
@@ -1,3 +1,127 @@
...
+
+You can read more details about the Docker Architecture [here](https://docs.docker.com/get-started/overview/#docker-architecture).
+
+We'll cover **Registry** on the core concept below.
+
+### Docker Installation
+
+You can install docker directly from their website at [https://docs.docker.com/get-docker/](https://docs.docker.com/get-docker/).
+
+## Concepts of Docker
```
[the remark this answers is not in the document]
Comment (hinxcode) [reply fundamentals/computing-development-environments/docker.md]: Done!

File: fundamentals/computing-development-environments/docker.md [hunk 9]
```diff
@@ -1,3 +1,127 @@
...
+- **Isolation**: Docker containers encapsulate applications and their dependencies, ensuring that they run consistently and independently of the host environment.
+- **Portability**: Docker containers can be easily moved between different environment: development, testing, and production. This guarantees consistent behavior.
+- **Version Control**: Docker file and/or images can be versioned, allowing you to maintain and track different versions of your application.
+- **Microservices Architecture**: Docker containers are well-suited for microservices-based architectures, enabling you to break down complex applications into smaller, manageable components.
+
+### Containers vs VM
+
+![container-vm](https://www.docker.com/wp-content/uploads/2021/11/docker-containerized-and-vm-transparent-bg.png)
+
+### Virtual Machine (VM)
```
[the remark this answers is not in the document]
Comment (hinxcode) [reply fundamentals/computing-development-environments/docker.md]: Fixed!

File: fundamentals/computing-development-environments/docker.md [hunk 10]
```diff
@@ -1,3 +1,127 @@
 # Docker
 
-Coming soon ([#34](https://github.com/uw-ssec/rse-guidelines/issues/34))
+## What is Docker?
```
[the remark this answers is not in the document]
Comment (hinxcode) [reply fundamentals/computing-development-environments/docker.md]: Fixed!

File: fundamentals/computing-development-environments/docker.md [hunk 11]
```diff
@@ -1,3 +1,127 @@
...
+
+Docker is a containerization platform that simplifies packaging applications and their dependencies into lightweight, portable containers.
+These containers can run consistently across different environments, eliminating the "it works on my machine" problem.
+
+**Key features**:
+
+- **Isolation**: Docker containers encapsulate applications and their dependencies, ensuring that they run consistently and independently of the host environment.
+- **Portability**: Docker containers can be easily moved between different environment: development, testing, and production. This guarantees consistent behavior.
```
[the remark this answers is not in the document]
Comment (hinxcode) [reply fundamentals/computing-development-environments/docker.md]: Done!

File: fundamentals/computing-development-environments/docker.md [hunk 12]
```diff
@@ -1,3 +1,127 @@
...
+- **Version Control**: Docker file and/or images can be versioned, allowing you to maintain and track different versions of your application.
```
[the remark this answers is not in the document]
Comment (hinxcode) [reply fundamentals/computing-development-environments/docker.md]: Fixed!

File: fundamentals/computing-development-environments/docker.md [hunk 13]
```diff
@@ -1,3 +1,102 @@
...
+
+#### Swarm
+
+Docker Swarm is an open-source container orchestration platform that is native to Docker. It converts multiple Docker instances into a single virtual host. Unlike Kubernetes' steeper learning curve, Swarm mode integrates seamlessly with Docker CLI, providing built-in service discovery, load balancing, and rolling updates with minimal configuration. For more details, see our [Docker Advanced page](../../advanced/deployment-strategies-tools/docker-advanced.html).
+
+#### Differences between Kubernetes and Swarm
+
+While both Docker Swarm and Kubernetes do manage containers across multiple hosts, the way they do it and the level of abstraction they provide differ. Docker Swarm's approach is more straightforward, while Kubernetes offers a more comprehensive but complex cluster management system. You can gain a better understanding of the distinction by reviewing the [Kubernetes](../../advanced/deployment-strategies-tools/kubernetes.html) and [Docker Advanced](../../advanced/deployment-strategies-tools/docker-advanced.html) pages.
```
Comment (nikiburggraf) [inline fundamentals/computing-development-environments/docker.md:102]: This is beautiful and very clear, great job!!

**Review by nikiburggraf — APPROVED** [review]
Looks incredible, thank you for making those changes!!

:shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit:

**Pull request merged**